gitops/argocd/image/register.sh

```diff
@@ -0,0 +1,134 @@
+	printf "insecured connection to Argo CD allowed!\n"
+        insecure="--insecure"
+    else
+        insecure=""
+    fi
+}
+
+# populate clusters with the cluster names taken from the kubconfig
+# populate kubconfigs with the associated kubeconfig for each cluster name
+get_clusters() {
+    clusters=()
+    kubeconfigs=()
+    # TODO: Romain has /credentials/kubeconfig/plnsvc
+    # to clarify: is it a directory plnsvc for all the workload clusters or is plnsvc already specific to 1 workload cluster?
+    # If it is the later we need to differentiate between workload clusters (a subdirectory?), kcp and Argo CD credentials
+    # something like:
+    #   credentials/kubeconfig/workload/plnsvc1
+    #   credentials/kubeconfig/workload/plnsvc2
+    #   credentials/kubeconfig/workload/plnsvc3
+    # If it the former, why plnsvc as name? How does it identify the role they are playing within the service?
+    # Looking at /environment it seems to be the former.
+    files=($(ls $DATA_DIR/gitops/credentials/kubeconfig/plnsvc))
+    for kubeconfig in ${files[@]}; do
+        clusters_sub=($(KUBECONFIG=${DATA_DIR}/gitops/credentials/kubeconfig/plnsvc/${kubeconfig} kubectl config get-clusters | cut -d ':' -f 1 ))
+        clusters_sub=( ${clusters_sub[@]:1} )
+        clusters+=( ${clusters_sub[@]} )
+        for i in "${!clusters_sub[@]}"; do
+            kubeconfigs+=( ${kubeconfig} )
+        done
+    done
+}
+
+parent_path=$( cd "$(dirname "${BASH_SOURCE[0]}")" ; pwd -P )
+
+prechecks
+
+printf "Retrieving clusters\n"
+get_clusters
+
+printf "Logging into Argo CD\n"
+# TODO: there may be a better way than user/password
+# there should be no assumption that Argo CD runs on the same cluster
+argocd ${insecure} login $ARGO_URL --username $ARGO_USER --password $ARGO_PWD
+
+printf "Getting the list of registered clusters\n"
+existing_clusters=$(argocd cluster list -o json | jq '.[].name')
+
+printf "Registering clusters and provisioning credentials for Argo CD\n"
+for i in "${!clusters[@]}"; do
+    printf "Processing cluster %s\n" "${clusters[$i]}"
+    if echo "${existing_clusters}" | grep "${clusters[$i]}"; then
+        printf "Cluster already registered\n"
+    else
+        printf "Registering cluster\n"
+	# TODO: Romain has /environment/plnsvc/cluster-name/config
+	# to clarify:
+	# - do we need "config". It seems I don't
+	# - as above is "plnsvc" the right name? workload or compute may be a better choice.
+	# - the directory name needs to match with the cluster name in kubeconfig
+	# - need to document that a directory /environment/plnsvc/${clusters[$i]}/ns-rbac with a kustomize having
+	# /environment/plnsvc/base/ns-rbac as a base needs to be created for every new cluster
+        # reapply rbac:
+	# - `argocd cluster add` requires the namespaces to exist, hence the first run of kubectl apply
+	# - `argocd cluster add` applies default rbac that may differ from what is desired, hence the second run
```

Comment:
If our intention is to just create the namespaces in the first run, we could create another directory for handling this better.

- pipelines-service/gitops/environment/plnsvc/base/
  - ns
    - kcp-syncer-ns.yaml
    - management-ns.yaml
    - tekton-ns.yaml
    - kustomization.yaml
  - rbac
    - argo-syncer-role.yaml
    - argo-tekton-role.yaml
    - argo-syncer-rolebinding.yaml
    - argo-tekton-rolebinding.yaml
    - kustomization.yaml

With this kind of a directory structure, lines 129 and 131 could call kustomization.yaml under ns and rbac respectively.

Reply:
Yes I thought of that. I was not sure of what would be better as it breaks a logical unit into peaces due to the way the Argo CD client handles things. I can do the split if you like ti better

Reply:
Hmm agreed. But since we are going the kustomize way, let's do it that way. It creates a neater way to add/remove stuff in the future too.
(I am doing something similar for "ArgoCD in kubernetes vs openshift" too.)